# **Manipulating Data with Pandas**

_Adapted from Zero to Mastery Machine Learning and Data Science Course [materials](https://github.com/mrdbourke/zero-to-mastery-ml/blob/master/section-2-data-science-and-ml-tools/introduction-to-pandas.ipynb)._

## What is pandas?

If you're getting into machine learning and data science and you're using Python, you're going to use pandas.

[pandas](https://pandas.pydata.org/) is an open source library which helps you analyse and manipulate data.

## Why pandas?

pandas provides a simple to use but very capable set of functions you can use to on your data.

It's integrated with many other data science and machine learning tools which use Python so having an understanding of it will be helpful as another tool in your toolbox.

One of the main use cases you'll come across is using pandas to transform your data in a way which makes it usable with machine learning algorithms.

## Importing pandas

To get started using pandas, the first step is to import it. 

The most common way (and method you should use) is to import pandas as the abbreviation `pd` (e.g. `pandas` -> `pd`).

If you see the letters `pd` used anywhere in machine learning or data science, it's probably referring to the pandas library.

The keyword `as` is used to give alias to imported modules. In this case, we're giving the alias `pd` to the `pandas` library.

In [52]:
# Using pd as alias for pandas
import pandas as pd

# Print the version
print(f"pandas version: {pd.__version__}")

pandas version: 3.0.2


## Datatypes

pandas has two main datatypes, `Series` and `DataFrame`.
* [`pandas.Series`](https://pandas.pydata.org/docs/reference/api/pandas.Series.html) - a 1-dimensional column of data.
* [`pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) (most common) - a 2-dimesional table of data with rows and columns.

You can create a `Series` using `pd.Series()` and passing it a Python list.

In [53]:
# Creating a series of car types
cars = pd.Series(["BMW", "Toyota", "Honda"])
cars

0       BMW
1    Toyota
2     Honda
dtype: str

In [54]:
# Creating a series of colours
colours = pd.Series(["Blue", "Red", "White"])
colours

0     Blue
1      Red
2    White
dtype: str

You can create a `DataFrame` by using `pd.DataFrame()` and passing it a Python dictionary.

Let's use our two `Series` as the values.

In [55]:
# Creating a DataFrame of cars and colours
car_data = pd.DataFrame({"Car type": cars, "Colour": colours})
car_data

,Car type,Colour
0,BMW,Blue
1,Toyota,Red
2,Honda,White


You can see the keys of the dictionary became the column headings (text in bold) and the values of the two `Series`'s became the values in the DataFrame.

It's important to note, many different types of data could go into the DataFrame. 

Here we've used only text but you could use floats, integers, dates and more.

In [56]:
# Creating a DataFrame using a dictionary
employees = pd.DataFrame(
    {
        "name": ["Alice", "Bob", "Charlie"],
        "age": [25, 30, 35],
        "department": ["HR", "Engineering", "Marketing"],
    }
)
employees

,name,age,department
0,Alice,25,HR
1,Bob,30,Engineering
2,Charlie,35,Marketing


In [57]:
# Creating a DataFrame using list of dictionaries
employees = pd.DataFrame(
    [
        {"name": "Alice", "age": 25, "department": "HR"},
        {"name": "Bob", "age": 30, "department": "Engineering"},
        {"name": "Charlie", "age": 35, "department": "Marketing"},
    ]
)
employees

,name,age,department
0,Alice,25,HR
1,Bob,30,Engineering
2,Charlie,35,Marketing


## Importing data

Creating `Series` and `DataFrame`'s from scratch is nice but what you'll usually be doing is importing your data in the form of a `.csv` (comma separated value), spreadsheet file or something similar such as an [SQL database](https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html).

pandas allows for easy importing of data like this through functions such as [`pd.read_csv()`](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) and [`pd.read_excel()`](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html) (for Microsoft Excel files).

Say you wanted to get this information from this Google Sheet document into a pandas `DataFrame`. You could export it as a `.csv` file and then import it using `pd.read_csv()`. Its also possible to import it directly from the URL of the Google Sheet document. 

> **Tip:** If the Google Sheet is public, `pd.read_csv()` can read it via URL, try searching for "pandas read Google Sheet with URL".

In [ ]:
# https://github.com/pandas-dev/pandas/blob/main/pandas/tests/io/data/csv/tips.csv
# Import tips data
tips_dataset = "../dataset/tips.csv"

# Option 1: Read from a local file
tips = pd.read_csv(tips_dataset)  # takes a filename as string as input
tips

# Option 2: Read directly from a URL/Google Sheets
# tips = pd.read_csv("...")
# tips

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


Now we've got the same data from the spreadsheet available in a pandas `DataFrame` called `tips`. 

Having your data available in a `DataFrame` allows you to take advantage of all of pandas functionality on it.

Another common practice you'll see is data being imported to `DataFrame` called `df` (short for `DataFrame`).

In [59]:
df = pd.read_csv(tips_dataset)
df

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


This `DataFrame` now holds the data from the spreadsheet and we can use all of pandas functionality on it.

Some common functions you might want to use on your `DataFrame` include:
* `DataFrame.head()` - shows the first 5 rows of the DataFrame.
* `DataFrame.tail()` - shows the last 5 rows of the DataFrame.
* `DataFrame.info()` - shows a summary of the DataFrame including the datatypes of each column and how many non-null values there are.
* `DataFrame.describe()` - shows a statistical summary of the DataFrame (only works on numeric columns).

This will be explored further in the next section.

## Exporting data

After you've made a few changes to your data, you might want to export it and save it so someone else can access the changes.

pandas allows you to export `DataFrame`'s to `.csv` format using [`.to_csv()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html) or spreadsheet format using [`.to_excel()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_excel.html).

We haven't made any changes yet to the `DataFrame` but let's try export it.

Adding `index=False` to the `to_csv()` function prevents pandas from adding an extra column with the index values (0, 1, 2, etc.) to the exported file.

In [ ]:
exported_csv_path = "../dataset/exported_tips.csv"

# exports the DataFrame to a CSV file without the index column
# df.to_csv(exported_csv_path, index=False)

## Describing data

One of the first things you'll want to do after you import some data into a pandas `DataFrame` is to start exploring it.

pandas has many built in functions which allow you to quickly get information about a `DataFrame`.

Let's explore some using the `tips` `DataFrame`.

In [61]:
tips

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


[`.dtypes`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dtypes.html) shows us what datatype each column contains.

In [62]:
tips.dtypes

total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object

[`.describe()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) gives you a quick statistical overview of the numerical columns.

In [63]:
tips.describe()

,total_bill,tip,size
count,244.000000,244.000000,244.000000
mean,19.785943,2.998279,2.569672
std,8.902412,1.383638,0.951100
min,3.070000,1.000000,1.000000
25%,13.347500,2.000000,2.000000
50%,17.795000,2.900000,2.000000
75%,24.127500,3.562500,3.000000
max,50.810000,10.000000,6.000000


[`.info()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html) shows a handful of useful information about a `DataFrame` such as: 
* How many entries (rows) there are 
* Whether there are missing values (if a columns non-null value is less than the number of entries, it has missing values)
* The datatypes of each column

In [64]:
tips.info()

<class 'pandas.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    str    
 3   smoker      244 non-null    str    
 4   day         244 non-null    str    
 5   time        244 non-null    str    
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), str(4)
memory usage: 13.5 KB


You can also call various statistical and mathematical methods such as [`.mean()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html) or [`.sum()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) directly on a `DataFrame` or `Series`.

Adding paramter `numeric_only=True` to the `mean` function tells pandas to only calculate the mean of the numeric columns and ignore any non-numeric columns, as calculating the mean of non-numeric columns doesn't make sense and would result in an error.

In [65]:
tips.mean(numeric_only=True)

total_bill    19.785943
tip            2.998279
size           2.569672
dtype: float64

In [66]:
ages = pd.Series([25, 30, 35])
ages.mean()

np.float64(30.0)

In [67]:
# Sum of all numeric columns, ignoring non-numeric columns
tips.sum(numeric_only=False)

total_bill                                              4827.77
tip                                                      731.58
sex           FemaleMaleMaleMaleFemaleMaleMaleMaleMaleMaleMa...
smoker        NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
day           SunSunSunSunSunSunSunSunSunSunSunSunSunSunSunS...
time          DinnerDinnerDinnerDinnerDinnerDinnerDinnerDinn...
size                                                        627
dtype: object

In [68]:
tips.sum(numeric_only=True)

total_bill    4827.77
tip            731.58
size           627.00
dtype: float64

In [69]:
# Calling sum to pandas Series
ages.sum()

np.int64(90)

Calling these on a whole `DataFrame` may not be as helpful as targeting an individual column. But it's helpful to know they're there.

`.columns` will show you all the columns of a `DataFrame`.

In [70]:
tips.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='str')

You can save them as a list which you could use later. Calling `tolist()` on pandas object returns a python list of the values. 

In [71]:
# Save tips columns  list
tips_columns = tips.columns.tolist()
tips_columns[0]

'total_bill'

`.index` will show you the values in a `DataFrame`'s index (the column on the far left).

pandas `DataFrame`'s, like Python lists, are 0-indexed (unless otherwise changed). This means they start at 0. 

In [72]:
tips.index

RangeIndex(start=0, stop=244, step=1)

In [73]:
# Show the length of the DataFrame; this also mean the number of rows in the DataFrame
len(tips)

244

So even though the length of our `tips` dataframe is 244, this means the indexes go from 0-243.

In [74]:
print("Smallest index:", tips.index.min())
print("Largest index:", tips.index.max())

Smallest index: 0
Largest index: 243


## 5. Viewing and selecting data

Some common methods for viewing and selecting data in a pandas DataFrame include:

* [`DataFrame.head(n=5)`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) - Displays the first `n` rows of a DataFrame (e.g. `tips.head()` will show the first 5 rows of the `car_sales` DataFrame).
* [`DataFrame.tail(n=5)`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html) - Displays the last `n` rows of a DataFrame.
* [`DataFrame.loc[]`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html) - Accesses a group of rows and columns by labels or a boolean array.
* [`DataFrame.iloc[]`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html) - Accesses a group of rows and columns by integer indices (e.g. `tips.iloc[0]` shows all the columns from index `0`.)
* [`DataFrame.columns`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.columns.html) -  Lists the column labels of the DataFrame.
* `DataFrame['A']` - Selects the column named `'A'` from the DataFrame.
* `DataFrame[DataFrame['A'] > 5]` - Boolean indexing filters rows based on column values meeting a condition (e.g. all rows from column `'A'` greater than `5`.

In practice, you'll constantly be making changes to your data, and viewing it. Changing it, viewing it, changing it, viewing it.

You won't always want to change all of the data in your `DataFrame`'s either. So there are just as many different ways to select data as there is to view it.

`.head()` allows you to view the first 5 rows of your `DataFrame`. You'll likely be using this one a lot. 

In [75]:
# Show the first 5 rows of the DataFrame
tips.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


Want more than 5? Just pass the value of how many rows you want to see as the parameter.

In [76]:
tips.head(10)

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
5,25.29,4.71,Male,No,Sun,Dinner,4
6,8.77,2.00,Male,No,Sun,Dinner,2
7,26.88,3.12,Male,No,Sun,Dinner,4
8,15.04,1.96,Male,No,Sun,Dinner,2
9,14.78,3.23,Male,No,Sun,Dinner,2


`.tail()` allows you to see the bottom 5 rows of your `DataFrame`. This is helpful if your changes are influencing the bottom rows of your data.

In [77]:
# Show the last 5 rows of the DataFrame
tips.tail()

,total_bill,tip,sex,smoker,day,time,size
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2
243,18.78,3.00,Female,No,Thur,Dinner,2


You can use `.loc[]` and `.iloc[]` to select data from your `Series` and `DataFrame`'s.

In [78]:
# Create a sample series
animals = pd.Series(
    ["cat", "dog", "bird", "snake", "ox", "lion"], index=[0, 3, 9, 8, 67, 3]
)
animals

0       cat
3       dog
9      bird
8     snake
67       ox
3      lion
dtype: str

In [79]:
# Select all indexes with 3
animals.loc[3]

3     dog
3    lion
dtype: str

In [80]:
# Select index 9
animals.loc[9]

'bird'

Let's try it out on our `tips` `DataFrame`.

In [81]:
tips

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


In [82]:
# Select row at index 3
tips.loc[3]

total_bill     23.68
tip             3.31
sex             Male
smoker            No
day              Sun
time          Dinner
size               2
Name: 3, dtype: object

[`iloc[]`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html) does a similar thing but works with exact positions.

In [83]:
animals

0       cat
3       dog
9      bird
8     snake
67       ox
3      lion
dtype: str

In [84]:
# Select row at position 3
animals.iloc[3]

'snake'

Even though `'snake'` appears at index 8 in the series, it's shown using `.iloc[3]` because it's at the 3rd (starting from 0) position.

Let's try with the `tips` `DataFrame`.

In [85]:
# Select row at position 3
tips.iloc[3]

total_bill     23.68
tip             3.31
sex             Male
smoker            No
day              Sun
time          Dinner
size               2
Name: 3, dtype: object

You can see it's the same as `.loc[]` because the index is in order, position 3 is the same as index 3.

You can also use slicing with `.loc[]` and `.iloc[]`.

In [86]:
# Get all rows up to position 3
animals.iloc[:3]

0     cat
3     dog
9    bird
dtype: str

In [87]:
# Get all rows up to (and including) index 3
tips.loc[:3]

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2


In [88]:
# Get all rows of the "day" column
tips.loc[:, "day"]  # note: ":" stands for "all", e.g. "all indices in the first axis"

0       Sun
1       Sun
2       Sun
3       Sun
4       Sun
       ... 
239     Sat
240     Sat
241     Sat
242     Sat
243    Thur
Name: day, Length: 244, dtype: str

When should you use `.loc[]` or `.iloc[]`?
* Use `.loc[]` when you're selecting rows and columns **based on their lables or a condition** (e.g. retrieving data for specific columns).
* Use `.iloc[]` when you're selecting rows and columns **based on their integer index positions** (e.g. extracting the first ten rows regardless of the labels).

However, in saying this, it will often take a bit of practice with each of the methods before you figure out which you'd like to use.

If you want to select a particular column, you can use `DataFrame.['COLUMN_NAME']`.

In [89]:
# Select tip column
tips["tip"]

0      1.01
1      1.66
2      3.50
3      3.31
4      3.61
       ... 
239    5.92
240    2.00
241    2.00
242    1.75
243    3.00
Name: tip, Length: 244, dtype: float64

In [90]:
# Select record where day is "Sat"
tips[tips["day"] == "Sat"]

,total_bill,tip,sex,smoker,day,time,size
19,20.65,3.35,Male,No,Sat,Dinner,3
20,17.92,4.08,Male,No,Sat,Dinner,2
21,20.29,2.75,Female,No,Sat,Dinner,2
22,15.77,2.23,Female,No,Sat,Dinner,2
23,39.42,7.58,Male,No,Sat,Dinner,4
...,...,...,...,...,...,...,...
238,35.83,4.67,Female,No,Sat,Dinner,3
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2


If you want to compare more columns in the context of another column, you can use `.groupby()`.

In [91]:
# Group by the time column and find the mean of the other columns
tips.groupby(["time"]).mean(numeric_only=True)

,total_bill,tip,size
time,,,
Dinner,20.797159,3.102670,2.630682
Lunch,17.168676,2.728088,2.411765


GroupBy also possible to use multiple columns as well. If multiple columns are passed to the `groupby()` function, it will group the data based on the unique combinations of values in those columns. For example, if you group by both "time" and "day", it will create groups for each unique combination of time and day (e.g., "Lunch on Saturday", "Dinner on Sunday", etc.). Then, when you apply an aggregation function like `mean()`, it will calculate the mean for each of those groups separately.

In [92]:
# Group by the time and day columns and find the mean of the other columns
tips.groupby(["time", "day"]).mean(numeric_only=True)

total_bill       tip      size
time   day                                 
Dinner Fri    19.663333  2.940000  2.166667
       Sat    20.441379  2.993103  2.517241
       Sun    21.410000  3.255132  2.842105
       Thur   18.780000  3.000000  2.000000
Lunch  Fri    12.845714  2.382857  2.000000
       Thur   17.664754  2.767705  2.459016

## Manipulating data

You've seen an example of one way to manipulate data but pandas has many more. 

Let's start with string methods. Because pandas is based on Python, however you can manipulate strings in Python, you can do the same in pandas.

You can access the string value of a column using `.str`. 

In [93]:
# Lower the time column
tips["time"].str.lower()

0      dinner
1      dinner
2      dinner
3      dinner
4      dinner
        ...  
239    dinner
240    dinner
241    dinner
242    dinner
243    dinner
Name: time, Length: 244, dtype: str

Notice how it doesn't change the values of the original `tips` `DataFrame` unless we set it equal to.

In [94]:
# The time columns values is not lowered
tips.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


In [95]:
# create a new column with the lowercased time values
tips["time_lower"] = tips["time"].str.lower()
tips.head()

,total_bill,tip,sex,smoker,day,time,size,time_lower
0,16.99,1.01,Female,No,Sun,Dinner,2,dinner
1,10.34,1.66,Male,No,Sun,Dinner,3,dinner
2,21.01,3.50,Male,No,Sun,Dinner,3,dinner
3,23.68,3.31,Male,No,Sun,Dinner,2,dinner
4,24.59,3.61,Female,No,Sun,Dinner,4,dinner


Reassigning the column changes it in the original `DataFrame`. This trend occurs throughout all kinds of data manipulation with pandas.

Some functions have a parameter called `inplace` which means a `DataFrame` is updated in place without having to reassign it.

Let's see what it looks like in combination with `.fillna()`, a function which fills missing data. But the thing is, our table isn't missing any data.

In practice, it's likely you'll work with datasets which aren't complete. What this means is you'll have to decide whether how to fill the missing data or remove the rows which have data missing.

Let's check out what a version of our `tips` `DataFrame` might look like with missing values.

In [ ]:
tips_missing_dataset = "../dataset/tips_missing_data.csv"
tips_missing = pd.read_csv(tips_missing_dataset)
tips_missing

,total_bill,tip,sex,smoker,day,time,size
0,NaN,1.01,NaN,No,Sun,Dinner,2.0
1,10.34,1.66,Male,No,NaN,Dinner,3.0
2,21.01,3.50,NaN,No,Sun,Dinner,3.0
3,23.68,3.31,Male,No,Sun,Dinner,2.0
4,24.59,3.61,Female,No,Sun,NaN,4.0
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,NaN,Dinner,3.0
240,27.18,NaN,Female,Yes,Sat,NaN,2.0
241,22.67,2.00,Male,Yes,Sat,Dinner,2.0
242,NaN,1.75,Male,No,Sat,Dinner,2.0


Missing values are shown by `NaN` in pandas. This can be considered the equivalent of `None` in Python.

Let's use the [`.fillna()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html) function to fill the `tip` column with the average of the other values in the same column.

In [97]:
# Fill tip column missing values with mean
# inplace is set to False by default
tips_missing["tip"].fillna(tips_missing["tip"].mean(), inplace=False)

0      1.010000
1      1.660000
2      3.500000
3      3.310000
4      3.610000
         ...   
239    5.920000
240    2.969136
241    2.000000
242    1.750000
243    3.000000
Name: tip, Length: 244, dtype: float64

Now let's check the original `tips_missing` `DataFrame`.

In [98]:
tips_missing

,total_bill,tip,sex,smoker,day,time,size
0,NaN,1.01,NaN,No,Sun,Dinner,2.0
1,10.34,1.66,Male,No,NaN,Dinner,3.0
2,21.01,3.50,NaN,No,Sun,Dinner,3.0
3,23.68,3.31,Male,No,Sun,Dinner,2.0
4,24.59,3.61,Female,No,Sun,NaN,4.0
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,NaN,Dinner,3.0
240,27.18,NaN,Female,Yes,Sat,NaN,2.0
241,22.67,2.00,Male,Yes,Sat,Dinner,2.0
242,NaN,1.75,Male,No,Sat,Dinner,2.0


Because `inplace` is set to `False` (default), there's still missing values in the `tip` column. 

Instead of using `inplace`, let's resassign the column to the filled version.

We'll use the syntax `df[col] = df[col].fillna(value)` to fill the missing values in the `tip` column with the average of the other values in the same column.

In [99]:
tips_missing["tip"] = tips_missing["tip"].fillna(tips_missing["tip"].mean())

Now lets check the `tips_missing` `DataFrame` again. Notice how the missing values in the `tip` column have been filled with the average of the other values in the same column.

To check if `Series` or `DataFrame` has missing values, you can use the [`.isna()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isna.html) method. This will return a boolean `Series` or `DataFrame` of the same shape indicating where the missing values are located (with `True` for missing values and `False` for non-missing values).

In [ ]:
tips_missing[tips_missing["tip"].isna()]  # should show empty DataFrame

,total_bill,tip,sex,smoker,day,time,size


We've filled some values but there's still missing values in `tips_missing`. Let's say you wanted to remove any rows which had missing data and only work with rows which had complete coverage.

You can do this using [`.dropna()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html).

In [101]:
# Remove missing data
tips_missing.dropna()

,total_bill,tip,sex,smoker,day,time,size
3,23.68,3.310000,Male,No,Sun,Dinner,2.0
7,26.88,3.120000,Male,No,Sun,Dinner,4.0
9,14.78,2.969136,Male,No,Sun,Dinner,2.0
10,10.27,1.710000,Male,No,Sun,Dinner,2.0
12,15.42,1.570000,Male,No,Sun,Dinner,2.0
...,...,...,...,...,...,...,...
235,10.07,1.250000,Male,No,Sat,Dinner,2.0
236,12.60,1.000000,Male,Yes,Sat,Dinner,2.0
237,32.83,1.170000,Male,Yes,Sat,Dinner,2.0
238,35.83,4.670000,Female,No,Sat,Dinner,3.0


In [102]:
tips_missing

,total_bill,tip,sex,smoker,day,time,size
0,NaN,1.010000,NaN,No,Sun,Dinner,2.0
1,10.34,1.660000,Male,No,NaN,Dinner,3.0
2,21.01,3.500000,NaN,No,Sun,Dinner,3.0
3,23.68,3.310000,Male,No,Sun,Dinner,2.0
4,24.59,3.610000,Female,No,Sun,NaN,4.0
...,...,...,...,...,...,...,...
239,29.03,5.920000,Male,No,NaN,Dinner,3.0
240,27.18,2.969136,Female,Yes,Sat,NaN,2.0
241,22.67,2.000000,Male,Yes,Sat,Dinner,2.0
242,NaN,1.750000,Male,No,Sat,Dinner,2.0


They're still there..?

It's because `.dropna()` has `inplace=False` as default. We can either set `inplace=True` or reassign the `tips_missing` `DataFrame`.

In [ ]:
# The following two lines do the same thing
tips_missing.dropna(inplace=True)  # Operation happens inplace without reassignment
tips_missing = (
    tips_missing.dropna()
)  # tips_missing gets reassigned to same DataFrame but with dropped values

In [104]:
tips_missing

,total_bill,tip,sex,smoker,day,time,size
3,23.68,3.310000,Male,No,Sun,Dinner,2.0
7,26.88,3.120000,Male,No,Sun,Dinner,4.0
9,14.78,2.969136,Male,No,Sun,Dinner,2.0
10,10.27,1.710000,Male,No,Sun,Dinner,2.0
12,15.42,1.570000,Male,No,Sun,Dinner,2.0
...,...,...,...,...,...,...,...
235,10.07,1.250000,Male,No,Sat,Dinner,2.0
236,12.60,1.000000,Male,Yes,Sat,Dinner,2.0
237,32.83,1.170000,Male,Yes,Sat,Dinner,2.0
238,35.83,4.670000,Female,No,Sat,Dinner,3.0


Instead of removing or filling data, what if you wanted to create it?

For example, creating a column called `Status` for number of seats.

pandas allows for simple extra column creation on `DataFrame`'s. 

Three common ways are: 
1. Adding a `pandas.Series` as a column.
2. Adding a Python list as a column.
3. By using existing columns to create a new column. (like previous `time_lower` example)

In [105]:
import random

status_label = ["Processed", "Pending"]
# use random.choice to randomly select a status label for each row in the tips DataFrame
status = [random.choice(status_label) for _ in range(len(tips))]

# Column from list
tips["Status"] = status
tips.head()

,total_bill,tip,sex,smoker,day,time,size,time_lower,Status
0,16.99,1.01,Female,No,Sun,Dinner,2,dinner,Processed
1,10.34,1.66,Male,No,Sun,Dinner,3,dinner,Processed
2,21.01,3.50,Male,No,Sun,Dinner,3,dinner,Pending
3,23.68,3.31,Male,No,Sun,Dinner,2,dinner,Processed
4,24.59,3.61,Female,No,Sun,Dinner,4,dinner,Processed


In [106]:
# Column from other column
tips["total_bill_with_tip"] = tips["total_bill"] + tips["tip"]
tips.head()

,total_bill,tip,sex,smoker,day,time,size,time_lower,Status,total_bill_with_tip
0,16.99,1.01,Female,No,Sun,Dinner,2,dinner,Processed,18.00
1,10.34,1.66,Male,No,Sun,Dinner,3,dinner,Processed,12.00
2,21.01,3.50,Male,No,Sun,Dinner,3,dinner,Pending,24.51
3,23.68,3.31,Male,No,Sun,Dinner,2,dinner,Processed,26.99
4,24.59,3.61,Female,No,Sun,Dinner,4,dinner,Processed,28.20


This kind of column creation is called [**feature engineering**](https://en.wikipedia.org/wiki/Feature_engineering), the practice of enriching your dataset with more information (either from it directly or elsewhere).

As for column creation, you can also create a new column setting all values to a one standard value.

In [107]:
tips["store"] = "Main Street"
tips.head()

,total_bill,tip,sex,smoker,day,time,size,time_lower,Status,total_bill_with_tip,store
0,16.99,1.01,Female,No,Sun,Dinner,2,dinner,Processed,18.00,Main Street
1,10.34,1.66,Male,No,Sun,Dinner,3,dinner,Processed,12.00,Main Street
2,21.01,3.50,Male,No,Sun,Dinner,3,dinner,Pending,24.51,Main Street
3,23.68,3.31,Male,No,Sun,Dinner,2,dinner,Processed,26.99,Main Street
4,24.59,3.61,Female,No,Sun,Dinner,4,dinner,Processed,28.20,Main Street


Now there are additional columns in the `tips` `DataFrame` which we created using different methods. But after thinking about it, `total_bill_with_tip` is quite redundant.

You decide you better remove it to prevent confusion.

You can remove a column using [`.drop('COLUMN_NAME', axis=1)`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html).

In [ ]:
tips = tips.drop(columns=["total_bill_with_tip"], axis=1)
tips.head()

Why `axis=1`? Because that's the axis columns live on. Rows live on `axis=0`.

Let's say you wanted to shuffle the order of your `DataFrame` so you could split it into train, validation and test sets. And even though the order of your samples was random, you wanted to make sure.

To do so you could use [`.sample(frac=1)`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html). 

`.sample()` randomly samples different rows from a `DataFrame`. 

The `frac` parameter dictates the fraction, where 1 = 100% of rows, 0.5 = 50% of rows, 0.01 = 1% of rows.

You can also use `.sample(n=1)` where `n` is the number of rows to sample.

In [ ]:
tips_sampled = tips.sample(frac=0.5, random_state=42)  # Sample 50% of the data
tips_sampled

Notice how the rows remain intact but their order is mixed (check the indexes).

`.sample(frac=X)` is also helpful when you're working with a large `DataFrame`. 

Say you had 2,000,000 rows.

Running tests, analysis and machine learning algorithms on 2,000,000 rows could take a long time. And since being a data scientist or machine learning engineer is about reducing the time between experiments, you might begin with a sample of rows first.

For example, you could use `40k_rows = 2_mil_rows.sample(frac=0.05)` to work on 40,000 rows from a `DataFrame` called `2_mil_rows` containing 2,000,000 rows.

What if you wanted to get the indexes back in order?

You could do so using [`.reset_index()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html).

In [ ]:
# Reset the indexes of tips_sampled to be sequential starting from 0
tips_sampled.reset_index()

Calling `.reset_index()` on a `DataFrame` resets the index numbers to their defaults. It also creates a new `Index` column by default which contains the previous index values.

Finally, what if you wanted to apply a function to a column. Such as, converting the `smoker` column from `yes` and `no` to `True` and `False`.

You can do so using the [`.apply()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.apply.html) function and passing it a Python [lambda function](https://www.w3schools.com/python/python_lambda.asp).

In [ ]:
tips["smoker"] = tips["smoker"].apply(lambda x: x == "Yes")
tips

Then if we want to get only the smoker tables, we can use boolean indexing.

In [ ]:
# as smoker values are now boolean, we can use them to filter the DataFrame
tips_smoker = tips[tips["smoker"]].reset_index(drop=True)
tips_smoker